Documentation/man/imgfac.1
.TH IMGFAC 1 "April 2011" "imgfac"

.SH NAME
imgfac - Create and customize OS images for private and public clouds.
This tool can be launched as a daemon to provide a QMFv2 interface that other applications can use for building images and instantiating these images on supported clouds.  Alternatively, a single image can be built or instantiated using the command line options described below without creating a daemon process or using QMF.

Currently supported public cloud services are: Amazon EC2, Rackspace.
Currently supported private cloud services: RHEV-M, KVM.

.SH SYNOPSIS
.B imgfac [-h] [--version] [-v] [--debug] [--foreground] [--config CONFIG]
.B          [imgdir IMGDIR] [--timeout TIMEOUT] [--qmf] [--qpidd QPIDD]
.B          [--image IMAGE]
.B          [--template TEMPLATE] [--target TARGET] [--target TARGET]...
.B          [--credentials CREDENTIALS] [--provider PROVIDER] [--provider PROVIDER]...
.B          [--warehouse WAREHOUSE]
.B          [--image_bucket IMAGE_BUCKET]
.B          [--build_bucket BUILD_BUCKET]
.B          [--target_bucket TARGET_BUCKET]
.B          [--template_bucket TEMPLATE_BUCKET]
.B          [--icicle_bucket ICICLE_BUCKET]
.B          [--provider_bucket PROVIDER_BUCKET]

.SH OPTIONS

.B -h, --help 
Display the help page.

.B --version
imgfac version info.

.B -v, --verbose
Set verbose logging.

.B --debug
Set really verbose logging for debugging.

.B --config CONFIG
Image factory configuration file to use.  (default: /etc/imagefactory.conf)

.B --imgdir IMGDIR
Build image files in location specified. After the image is finished it will be saved in this location.  (default: /tmp)

.B --image IMAGE
The UUID of the image to build or push. This is a UUID of an object in IWHD's image bucket.

.B --timeout TIMEOUT
Set the timeout period for image building in seconds.  (default: 3600)

.SS QMF agent:
  Provide a QMFv2 agent interface.

.B --foreground
Stay in the foreground and avoid launching a daemon. 
Launching in foreground allows the user to see what messages are being passed to QMF. For more detailed output use this parameter together with .B --debug.
If the parameter is not specified, the QMF agent will be started as a daemon.
(default: False)

.B --qmf
Turn on QMF agent interface.  (default: False)

.B --qpidd QPIDD
URL of qpidd to connect to.  (default: localhost)

.SS Image building:
  Build specified system and exit.

.B --template TEMPLATE
Template to build from. Template can be passed as URL, PATH, UUID or XML. Please refer to the 
.BR XML
.BR EXAMPLES
section at the bottom.

.B --target TARGET
The list of cloud services to target. Examples include ec2, rhevm, vsphere, rackspace, condorcloud and gogrid.


.SS Image pushing:
  Push an image and exit.


.B --provider PROVIDER
The list of cloud service providers to push the emacs. Examples include ec2-us-east-1, ec2-us-west-1, rackspace.

.B --credentials CREDENTIALS
Cloud provider credentials. Credentials must be passed as XML file. Please refer to the 
.BR XML 
.BR EXAMPLES 
section at the bottom.

.SS Image Warehouse:
  Options for specifying Image Warehouse base URL and bucket names.

.B --warehouse WAREHOUSE
URL of the warehouse location to store images.  (default: http://localhost:9090/)

.B --image_bucket IMAGE_BUCKET
Name of warehouse bucket to look in for images.  (default: None)

.B --build_bucket BUILD_BUCKET
Name of warehouse bucket to look in for builds.  (default: None)

.B --target_bucket TARGET_BUCKET
Name of warehouse bucket to look in for target images.  (default: None)

.B --template_bucket TEMPLATE_BUCKET
Name of warehouse bucket to look in for templates.  (default: None)

.B --icicle_bucket ICICLE_BUCKET
Name of warehouse bucket to look in for icicles.  (default: None)

.B --provider_bucket PROVIDER_BUCKET
Name of warehouse bucket to look in for provider image instances. (default: None)

.SH XML EXAMPLES
.sp
.SS CREDENTIALS

It's assumed that you will have more credentials in the same XML file, so the syntax will be .B cloud_provider = [XML string]
eg.

rackspace = "<provider_credentials> \\
        <rackspace_credentials> \\
            <username>foo</username> \\
            <api_key>0123456789</api_key> \\
        </rackspace_credentials> \\
    </provider_credentials>"

.SS TEMPLATE 
eg. a Windows 2008 64bit example

<template>
  <name>win2kjeos</name>
  <os>
    <name>Windows</name>
    <version>2008</version>
    <arch>x86_64</arch>
    <install type='iso'>
      <iso>http://directory_path/windows2008x64.iso</iso>
    </install>
    <key></key>
  </os>
  <description>Windows 2008</description>
  <repos>
     <repo url="smb://domain\\user:UserName@IP\\share" name="Default"/>
     <repo url="smb://domain\\user:UserName@IP\\another_share" name="Alternative"/>
  </repos>
  <packages>
    <package>
        <repo name="Default"/>
        <name>Dot Net 4.0</name>
        <file>dotNetFx40_Full_x86_x64.exe</file>
        <arguments>/passive</arguments>
    </package>
    <package>
        <repo name="Alternative"/>
        <name>Winrar</name>
        <file>winrarx64393.exe</file>
        <arguments>/s</arguments>
    </package>
  </packages>
</template>






imgfac.py
#!/usr/bin/env python
# encoding: utf-8

#   Copyright 2011 Red Hat, Inc.
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.

import sys
import os
import os.path
import signal
import logging
from imagefactory.ApplicationConfiguration import ApplicationConfiguration
from imagefactory.qmfagent.ImageFactoryAgent import *
from imagefactory.BuildDispatcher import BuildDispatcher
from imagefactory.ImageWarehouse import ImageWarehouse
from imagefactory.Singleton import Singleton

class Application(Singleton):

    def qmf_agent():
        doc = "The qmf_agent property."
        def fget(self):
            return self._qmf_agent
        def fset(self, value):
            self._qmf_agent = value
        def fdel(self):
            del self._qmf_agent
        return locals()
    qmf_agent = property(**qmf_agent())

    def _singleton_init(self):
        super(Application, self)._singleton_init()
        self.log = logging.getLogger('%s.%s' % (__name__, self.__class__.__name__))
        self.daemon = False
        self.pid_file_path = "/var/run/imagefactory.pid"
        signal.signal(signal.SIGTERM, self.signal_handler)
        self.app_config = ApplicationConfiguration().configuration

        # by setting TMPDIR here we make sure that libguestfs
        # (imagefactory -> oz -> libguestfs) uses the temporary directory of
        # the user's choosing
        os.putenv('TMPDIR', self.app_config['tmpdir'])

    def __init__(self):
        pass

    def setup_logging(self):
        if (self.app_config['foreground']):
            logging.basicConfig(level=logging.WARNING, format='%(asctime)s %(levelname)s %(name)s pid(%(process)d) Message: %(message)s')
        else:
            logging.basicConfig(level=logging.WARNING, format='%(asctime)s %(levelname)s %(name)s pid(%(process)d) Message: %(message)s', filename='/var/log/imagefactory.log')
        if (self.app_config['debug']):
            logging.getLogger('').setLevel(logging.DEBUG)
        elif (self.app_config['verbose']):
            logging.getLogger('').setLevel(logging.INFO)

    def signal_handler(self, signum, stack):
        """docstring for sigterm_handler"""
        if (signum == signal.SIGTERM):
            logging.warn('caught signal SIGTERM, stopping...')

            if (self.qmf_agent):
                self.qmf_agent.shutdown()
                try:
                    os.remove(self.pid_file_path)
                except Exception, e:
                    self.log.warning(str(e))

            sys.exit(0)

    # TODO: (redmine 273) - add code here to set the user:group we're running as and drop privileges
    def daemonize(self): #based on Python recipe 278731
        UMASK = 0
        WORKING_DIRECTORY = '/'
        IO_REDIRECT = os.devnull

        try:
            pid = os.fork()
        except OSError, e:
            raise Exception, "%s [%d]" % (e.strerror, e.errno)

        if (pid == 0):
            os.setsid()
            signal.signal(signal.SIGHUP, signal.SIG_IGN)
            try:
                pid = os.fork()
            except OSError, e:
                raise Exception, "%s [%d]" % (e.strerror, e.errno)

            if (pid == 0):
                os.chdir(WORKING_DIRECTORY)
                os.umask(UMASK)
            else:
                os._exit(0)
        else:
            os._exit(0)

        for file_descriptor in range(0, 2): # close stdin, stdout, stderr
            try:
                os.close(file_descriptor)
            except OSError:
                pass # The file descriptor wasn't open to begin with, just ignore

        os.open(IO_REDIRECT, os.O_RDWR)
        os.dup2(0, 1)
        os.dup2(0, 2)

        return(True)

    def main(self):
        if (self.app_config['qmf']):
            if (not self.app_config['foreground']):
                self.daemon = self.daemonize()

            self.setup_logging()
            if(self.daemon):
                try:
                    with open(self.pid_file_path, "w") as pidfile:
                        pidfile.write("%s\n" % (str(os.getpid()), ))
                        pidfile.close()
                except Exception, e:
                    logging.warning(str(e))
                logging.info("Launched as daemon...")
            elif(self.app_config['foreground']):
                logging.info("Launching in foreground...")
            else:
                logging.warning("Failed to launch as daemon...")

            self.qmf_agent = ImageFactoryAgent(self.app_config['qpidd'])
            self.qmf_agent.run()

        else:
            self.app_config['foreground'] = True
            self.setup_logging()

            if self.app_config['target_image'] and self.app_config['target'] and self.app_config['provider']:
                if len(self.app_config['target']) != 1:
                    print("Expect a single target, but '%s' supplied" % (self.app_config['target'],))
                    sys.exit(1)
                if len(self.app_config['provider']) != 1:
                    print("Expect a single provider, but '%s' supplied" % (self.app_config['provider'],))
                    sys.exit(1)
                image_ids = BuildDispatcher().import_image(self.app_config['image'], None, self.app_config['target_image'], self.app_config['image_desc'], self.app_config['target'][0], self.app_config['provider'][0])
                print("Imported target image ID %s as image %s" % (self.app_config['target_image'], image_ids[0]))
            elif (self.app_config['template'] and self.app_config['target']) or (self.app_config['image'] and not self.app_config['provider']):
                dispatchers = BuildDispatcher().build_image_for_targets(self.app_config['image'], None, self.app_config['template'], self.app_config['target'])
                for dispatcher in dispatchers:
                    print("Building build %s of image %s to target %s" % (dispatcher.build_id, dispatcher.image_id, dispatcher.target))

            elif (self.app_config['image'] and self.app_config['provider'] and self.app_config['credentials']):
                credentials = self.app_config['credentials']
                if(os.path.exists(credentials)):
                    credentials_file = open(credentials, "r")
                    file_contents = credentials_file.read()
                    credentials_file.close()
                    credentials = file_contents

                if(not (("<provider_credentials>" in file_contents.lower()) and ("</provider_credentials>" in file_contents.lower()))):
                    print("Unexpected content or formatting of credentials...")
                    sys.exit(1)

                dispatchers = BuildDispatcher().push_image_to_providers(self.app_config['image'], None, self.app_config['provider'], credentials)
                for dispatcher in dispatchers:
                    print("Pushing build %s of image %s to provider %s" % (dispatcher.build_id, dispatcher.image_id, dispatcher.provider))

if __name__ == "__main__":
    application = Application()
    sys.exit(application.main())
